cmd/kube-apiserver/app/preflight/checks.go

```diff
@@ -0,0 +1,102 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package preflight
+
+import (
+	"net"
+	"net/url"
+	"time"
+	"errors"
+	"fmt"
+
+	"k8s.io/kubernetes/cmd/kube-apiserver/app/options"
+)
+
+const retryLimit = 60
+const retryInterval = 1 * time.Second
+
+type clock interface {
+	Sleep(time.Duration)
+	After(time.Duration) <-chan time.Time
+}
+
+type connectionTimer struct{}
+func (connectionTimer) Sleep(d time.Duration) {time.Sleep(d)}
+func (connectionTimer) After(d time.Duration) <-chan time.Time {return time.After(d)}
+
+type connection interface {
+	serverReachable(address string) bool
+}
+
+type etcdConnection struct {}
+
+func (etcdConnection) serverReachable(address string) bool {
+	if conn, err := net.Dial("tcp", address); err == nil {
+		defer conn.Close()
+		return true
+	}
+	return false
+}
+
+func checkEtcdServer(address string, foundEtcd chan struct{}, stop chan struct{}, timer clock, etcd connection) {
+	retries := 0
+	for retries < retryLimit {
+		retries += 1
+		select {
+		case <-stop:
+			return
+		default:
+		}
+		if etcd.serverReachable(address) {
+			select {
+			case foundEtcd <- struct{}{}:
+			default:
+			}
+		}
+		timer.Sleep(retryInterval)
+	}
+}
+
+func WaitForEtcd(serverList []string, timer clock, etcd connection) error {
+	foundEtcd := make(chan struct{}, 1)
```

Comment:
Actually, no go routines will block. In checkEtcdServer, the attempt to send on foundEtcd is in a select so is asynchronous. each routine (that would have blocked) will attempt to send, fail, and choose the default case (which is a no-op). Each one will then sleep and stop.